Clustering/GMM/src/main/java/org/tribuo/clustering/gmm/GMMTrainer.java

```diff
@@ -0,0 +1,617 @@
+                        if (a[j] instanceof DenseMatrix aMat && b[j] instanceof DenseMatrix bMat) {
+                            aMat.intersectAndAddInPlace(bMat);
+                        } else if (a[j] instanceof DenseVector aVec && b[j] instanceof DenseVector bVec) {
+                            aVec.intersectAndAddInPlace(bVec);
+                        } else {
+                            throw new IllegalStateException("Invalid types in reduce, expected both DenseMatrix or DenseVector, found " + a[j].getClass() + " and " + b[j].getClass());
+                        }
+                    }
+                };
+                if (parallel) {
+                    try {
+                        covariances = fjp.submit(() -> zipMStream.parallel().collect(zeroTensor, mStep, combineTensor)).get();
+                    } catch (InterruptedException | ExecutionException e) {
+                        throw new RuntimeException("Parallel execution failed", e);
+                    }
+                } else {
+                    covariances = zipMStream.collect(zeroTensor, mStep, combineTensor);
+                }
+
+                // renormalize mixing distribution
+                double mixingSum = newMixingDistribution.sum();
+                newMixingDistribution.scaleInPlace(1/mixingSum);
+                mixingDistribution.setElements(newMixingDistribution);
+
+                // compute precisions
+                switch (covarianceType) {
+                    case FULL -> {
+                        for (int j = 0; j < covariances.length; j++) {
+                            DenseMatrix covMat = (DenseMatrix) covariances[j];
+                            covMat.intersectAndAddInPlace(covarianceJitter);
+                            Optional<DenseMatrix.CholeskyFactorization> optFact = covMat.choleskyFactorization();
+                            if (optFact.isPresent()) {
+                                DenseMatrix.CholeskyFactorization fact = optFact.get();
+                                precision[j] = fact.inverse();
+                                determinant[j] = fact.determinant();
+                            } else {
+                                throw new IllegalStateException("Failed to invert covariance matrix, cholesky didn't complete.");
+                            }
+                        }
+                    }
+                    case DIAGONAL, SPHERICAL -> {
+                        for (int j = 0; j < covariances.length; j++) {
+                            DenseVector covVec = (DenseVector) covariances[j];
+                            covVec.intersectAndAddInPlace(covarianceJitter);
+                            DenseVector preVec = (DenseVector) precision[j];
+                            double tmp = 1;
+                            for (int k = 0; k < preVec.size(); k++) {
+                                double curVal = 1/Math.sqrt(covVec.get(k));
+                                preVec.set(k, curVal);
```

Comment:
I know `precision[j]` gets set by mutation, but it might be nice to assign it anyway or have a comment for explicitness.